core/math/a_star_grid_2d.h

```diff
@@ -0,0 +1,207 @@
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef A_STAR_GRID_2D_H
+#define A_STAR_GRID_2D_H
+
+#include "core/object/gdvirtual.gen.inc"
+#include "core/object/ref_counted.h"
+#include "core/object/script_language.h"
+#include "core/templates/list.h"
+#include "core/templates/oa_hash_map.h"
+
+class AStarGrid2D : public RefCounted {
+	GDCLASS(AStarGrid2D, RefCounted);
+
+public:
+	enum DiagonalMode {
+		DIAGONAL_MODE_ALWAYS,
+		DIAGONAL_MODE_NEVER,
+		DIAGONAL_MODE_AT_LEAST_ONE_WALKABLE,
+		DIAGONAL_MODE_ONLY_IF_NO_OBSTACLES,
+		DIAGONAL_MODE_MAX,
+	};
+
+	enum Heuristic {
+		HEURISTIC_EUCLIDEAN,
+		HEURISTIC_MANHATTAN,
+		HEURISTIC_OCTILE,
+		HEURISTIC_CHEBYSHEV,
+		HEURISTIC_MAX,
+	};
+
+private:
+	int64_t width = 0;
+	int64_t height = 0;
+
+	bool jumping_enabled = false;
+	DiagonalMode diagonal_mode = DIAGONAL_MODE_ALWAYS;
+	Heuristic default_heuristic = HEURISTIC_EUCLIDEAN;
+
+	enum Dir {
+		DIR_LEFT,
+		DIR_RIGHT,
+		DIR_TOP,
+		DIR_TOP_LEFT,
+		DIR_TOP_RIGHT,
+		DIR_BOTTOM,
+		DIR_BOTTOM_LEFT,
+		DIR_BOTTOM_RIGHT,
+	};
+
+	struct Point {
+		int64_t id = 0;
+
+		_FORCE_INLINE_ int64_t get_x(int64_t p_grid_width) { return id % p_grid_width; }
+		_FORCE_INLINE_ int64_t get_y(int64_t p_grid_width) { return id / p_grid_width; }
+
+		real_t weight_scale = 0;
+		bool enabled = true;
+		Vector2 pos;
+
+		Point *neighbours[8] = { nullptr };
+
+		// Used for pathfinding.
+		Point *prev_point = nullptr;
+		real_t g_score = 0;
+		real_t f_score = 0;
+		uint64_t open_pass = 0;
+		uint64_t closed_pass = 0;
+
+		Point() {}
+	};
+
+	struct JumpState {
+		int64_t from;
+		int64_t to;
+		int stage = 0;
+	};
+
+	struct SortPoints {
+		_FORCE_INLINE_ bool operator()(const Point *A, const Point *B) const { // Returns true when the Point A is worse than Point B.
+			if (A->f_score > B->f_score) {
+				return true;
+			} else if (A->f_score < B->f_score) {
+				return false;
+			} else {
+				return A->g_score < B->g_score; // If the f_costs are the same then prioritize the points that are further away from the start.
+			}
+		}
+	};
+
+	OAHashMap<int64_t, Point *> points = 0U;
```

Comment:
Using an OAHashMap kind of beats the purpose of it being fast as a grid. I would simply use:

`LocalVector<LocalVector<Point>>` as a proper array. This will be much, much faster and you index as `points[x][y]`.